src/jsp/com/cubrid/plcsql/compiler/ast/Unit.java

```diff
@@ -0,0 +1,157 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
+ * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ *
+ */
+
+package com.cubrid.plcsql.compiler.ast;
+
+import com.cubrid.plcsql.compiler.Misc;
+import java.sql.*;
+
+public class Unit implements AstNode {
+
+    public enum TargetKind {
+        FUNCTION,
+        PROCEDURE
+    };
+
+    public final TargetKind targetKind;
+    public final boolean autonomousTransaction;
+    public final boolean connectionRequired;
+    public final String importsStr;
+    public final DeclRoutine routine;
+
+    public Unit(
+            TargetKind targetKind,
+            boolean autonomousTransaction,
+            boolean connectionRequired,
+            String importsStr,
+            DeclRoutine routine) {
+
+        this.targetKind = targetKind;
+        this.autonomousTransaction = autonomousTransaction;
+        this.connectionRequired = connectionRequired;
+        this.importsStr = importsStr;
+        this.routine = routine;
+    }
+
+    @Override
+    public String toJavaCode() {
+
+        String strGetConn;
+        if (connectionRequired) {
+            strGetConn =
+                    tmplGetConn.replace(
+                            "%AUTONOMOUS-TRANSACTION%",
+                            autonomousTransaction ? "?autonomous_transaction=true" : "");
+        } else {
+            strGetConn = "// connection not required";
+        }
+
+        String strDecls =
+                routine.decls == null
+                        ? "// no declarations"
+                        : tmplDeclClass
+                                .replace("%BLOCK%", routine.name.toLowerCase())
+                                .replace(
+                                        "  %DECLARATIONS%",
+                                        Misc.indentLines(routine.decls.toJavaCode(), 1));
+        ;
+        String strParams =
+                routine.paramList == null
+                        ? "// no parameters"
+                        : routine.paramList.toJavaCode(",\n");
+
+        return tmplUnit.replace("%IMPORTS%", importsStr)
+                .replace("%CLASS-NAME%", getClassName())
+                .replace(
+                        "%RETURN-TYPE%",
+                        routine.retType == null ? "void" : routine.retType.toJavaCode())
+                .replace("%METHOD-NAME%", routine.name)
+                .replace("      %PARAMETERS%", Misc.indentLines(strParams, 3))
+                .replace("    %GET-CONNECTION%", Misc.indentLines(strGetConn, 2))
+                .replace("    %DECL-CLASS%", Misc.indentLines(strDecls, 2))
+                .replace("    %BODY%", Misc.indentLines(routine.body.toJavaCode(), 2));
+    }
```

Comment:
Java's replace() is really slow in Java 8. The performance issue is fixed in Java 9, But It could be hard to upgrade our target Java version. So to workaround, Mostly Apache Commons Lang's StringUtils.replace() is recommended.

Reply:
Thank you for the valuable information!
Maybe, we can also consider shipping PL/CSQL feature with JDK 9 or JDK 13.
This article (https://stackoverflow.com/questions/16228992/commons-lang-stringutils-replace-performance-vs-string-replace) says JDK 13 further enhanced the replace method.

Reply:
It is hard to determine whether to consider upgrading the JDK version now. Since Java version upgrades in the QA test environment should also be considered, Discussing with the QA team is required. Since we are already using one library from Apache Common Lang in this PR, I think it would be good to use Apache's StringUtil library too.

Reply:
OK. Let me handle this in another issue.

Reply:
Ok, please create a new issue in advance to track it later.

Reply:
CBRD-24568 was created to improve Java code generation performance